discojs/src/dataset/dataset.ts

```diff
@@ -57,9 +62,7 @@ export class Dataset<T> implements AsyncIterable<T> {
   chain(other: Dataset<T> | DatasetLike<T>): Dataset<T> {
     if (!(other instanceof Dataset)) other = new Dataset(other);
 
-    const self = {
-      [Symbol.asyncIterator]: () => this.#content(),
-    };
+    const self = { [Symbol.asyncIterator]: () => this[Symbol.asyncIterator]() };
```

Comment:
[`this` is a nightmare](https://github.com/Microsoft/TypeScript/wiki/'this'-in-TypeScript#what-is-this-in-javascript)

in here, it's because `function` actually defines a new scope so a new `this` (which isn't the case with arrow function for example). so we need to keep the `this` of `Dataset` around instead of using the one of the function. and that's the way I found to do so at the time (it's also very explicit, which helps with `this` handling).

that being told, now that I read a bit more on it, I changed it up to append `bind(this)` to the function to forward `Dataset`'s `this`. it does look nicer (wasn't able to pass `super` in `CachingDataset` tough, that would be magic for another day)